COM-89: Implement testing indicator 3a

tmvumbi2 opened the pull request:
Jira ticket: https://jembiprojects.jira.com/browse/COM-89

tmvumbi2 committed:
COM-89 Implement testing indicator 3a
Implements testing indicator 3a

tmvumbi2 committed:
Merge branch 'dev' into COM-89

# Conflicts:
#	reportssql/testing_report_util_functions.sql

lauravignoli committed:
Merge branch 'dev' into COM-89

# Conflicts:
#	reportssql/indicator_functions.sql

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,39 @@ WHERE
+    patient pat
+WHERE
+    patientAgeAtReportEndDateIsBetween(pat.patient_id, p_endDate, p_startAgeInMonths, p_endAgeInMonths) AND
+    patientHadAVirologicHIVTestDuringReportingPeriod(pat.patient_id, p_startDate, p_endDate) AND
```
@tmvumbi2  There's no need to check that the patient had a virologic HIV Test and that they had a positive virologic HIV Test don't you think? The second condition should be sufficient?

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,39 @@ WHERE
+    patientHadAPositiveVirologicHIVTestResultDuringReportingPeriod(pat.patient_id, p_startDate, p_endDate) AND
+    patientMostRecentVirologicHIVTestResultIsPositive(pat.patient_id) AND
+    patientWasNotEnrolledToHIVProgramBeforeVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
+    patientWasNotInitiatedToARVBeforeVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
```
@tmvumbi2 Can't we remove this function and use:

NOT patientHasStartedARVTreatmentBefore(p_patientId, dateOfVirologicHIVTest) AND

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,39 @@ WHERE
+    patientMostRecentVirologicHIVTestResultIsPositive(pat.patient_id) AND
+    patientWasNotEnrolledToHIVProgramBeforeVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
+    patientWasNotInitiatedToARVBeforeVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
+    patientNotTakingARVAtDateOfVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
```
@tmvumbi2 Can't we remove this function and use:

NOT patientWasOnARVTreatmentOrHasPickedUpADrugWithinReportingPeriod(p_patientId, dateOfVirologicHIVTest, dateOfVirologicHIVTest, 0)) AND

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,39 @@ WHERE
+    patientHadAVirologicHIVTestDuringReportingPeriod(pat.patient_id, p_startDate, p_endDate) AND
+    patientHadAPositiveVirologicHIVTestResultDuringReportingPeriod(pat.patient_id, p_startDate, p_endDate) AND
+    patientMostRecentVirologicHIVTestResultIsPositive(pat.patient_id) AND
+    patientWasNotEnrolledToHIVProgramBeforeVirologicTest(pat.patient_id, p_startDate, p_endDate) AND
```
@tmvumbi2 Can't we remove this function and use:

NOT patientHasEnrolledIntoHivProgramBefore(pat.patient_id,dateOfVirologicHIVTest) AND

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,39 @@ WHERE
+    DETERMINISTIC
+BEGIN
+    DECLARE result INT(11) DEFAULT 0;
+
```
@tmvumbi2 We can define dateOfVirologicHIVTest here:

DECLARE dateOfVirologicHIVTest DATE;
SET dateOfVirologicHIVTest = getDateOfVirologicTest(p_patientId, p_startDate, p_endDate);

lauravignoli commented on reportssql/testing_report_util_functions.sql:
```diff
@@ -458,6 +458,201 @@ BEGIN
+END$$
+DELIMITER ;
+
+-- getDateOfVirologicTest
```
@tmvumbi2 Now that this function is created, patientHadAVirologicHIVTestDuringReportingPeriod can be removed and dateOfVirologicHIVTest is not null can be used?

lauravignoli replied:
This applies for indicator 2

lauravignoli committed:
COM-89 Remove unnecessary functions

lauravignoli committed:
COM-89 Remove unnecessary check

There's no need to check that the patient had a virologic HIV Test and that they had a positive virologic HIV Test as well. The second condition should be sufficient.

lauravignoli commented on reportssql/testing_report_util_functions.sql:
```diff
@@ -458,6 +458,141 @@ BEGIN
+    WHERE o.voided = 0
+        AND o.person_id = p_patientId
+        AND c.uuid = pcrExamUuid
+        AND o.value_coded = (SELECT concept.concept_id FROM concept WHERE concept.uuid = positiveUuid)
```
@tmvumbi2 This function doesn't check that the most recent virologic test is positive, because it searches for positive tests only.

lauravignoli replied:
You could check if there's a negative test after getDateOfVirologicTest?

tmvumbi2 committed:
COM-89 Remove function patientHadAVirologicHIVTestDuringReportingPeriod
The above function can be replaced by getDateOfVirologicTest IS NOT NULL

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,42 @@ WHERE
+    NOT patientHasEnrolledIntoHivProgramBefore(pat.patient_id, dateOfVirologicHIVTest) AND
+    NOT patientHasStartedARVTreatmentBefore(p_patientId, dateOfVirologicHIVTest) AND
+    NOT patientWasOnARVTreatmentOrHasPickedUpADrugWithinReportingPeriod(p_patientId, dateOfVirologicHIVTest, dateOfVirologicHIVTest, 0) AND
+    patientAgeAtVirologicHIVTestIsBetween(pat.patient_id, p_startAgeInMonths, p_endAgeInMonths, p_startDate, p_endDate, p_includeStartAge) AND
```
@tmvumbi2 I don't see this condition in the spec?

lauravignoli commented on reportssql/indicator_functions.sql:
```diff
@@ -913,6 +913,42 @@ WHERE
+FROM
+    patient pat
+WHERE
+    patientAgeAtReportEndDateIsBetween(pat.patient_id, p_endDate, p_startAgeInMonths, p_endAgeInMonths) AND
```
@tmvumbi2 I don't think this check is necessary, and it's also wrong, because for dissagregation we should check the age at test date (as it's done later on in patientAgeAtVirologicHIVTestIsBetween) and not at the report end date.

I think we should remove this check and delete the function.

tmvumbi2 committed:
COM-89 Fix patientMostRecentVirologicHIVTestResultIsPositive
The previous implementation was incorrect as it was selecting only positive results (i.e. it couldn't return the correct result if the most recent test was negative)

tmvumbi2 committed:
COM-89 Remove function patientAgeAtReportEndDateIsBetween
The age check is already done by patientAgeAtVirologicHIVTestIsBetween

lauravignoli committed:
COM-89 Fix function

Use datetime instead of  date created to check obs, to cover for retrospective data entry.

lauravignoli commented on reportssql/testing_report_util_functions.sql:
```diff
@@ -458,6 +458,201 @@ BEGIN
+CREATE FUNCTION getDateOfVirologicTest(
+    p_patientId INT(11),
+    p_startDate DATE,
+    p_endDate DATE) RETURNS TINYINT(1)
```
@tmvumbi2 This function should return DATE, not TINYINT(1) right?

lauravignoli commented on reportssql/testing_report_util_functions.sql:
```diff
@@ -458,6 +458,201 @@ BEGIN
+    SET dateAtVirologicHIVTestLabForm = getDateOfVirologicHIVTestFromLabForm(p_patientId, p_startDate, p_endDate);
+    SET dateAtVirologicHIVTestElis = getDateOfVirologicHIVTestFromElis(p_patientId, p_startDate, p_endDate);
+
+    IF dateAtVirologicHIVTestLabForm IS NULL AND dateAtVirologicHIVTestElis IS NULL THEN
```
@tmvumbi2 Maybe this is a better way to write this:

    IF dateAtVirologicHIVTestElis IS NOT NULL THEN
        RETURN dateAtVirologicHIVTestElis;
    ELSE
        RETURN dateAtVirologicHIVTestLabForm;
    END IF;

lauravignoli commented on reportssql/testing_report_util_functions.sql:
```diff
@@ -458,6 +458,201 @@ BEGIN
+
+    SET dateOfVirologicHIVTest = getDateOfVirologicTest(p_patientId, p_startDate, p_endDate);
+
+    RETURN (NOT patientWasOnARVTreatmentOrHasPickedUpADrugWithinReportingPeriod(p_patientId, dateOfVirologicHIVTest, dateOfVirologicHIVTest, 0));
```
@tmvumbi2 This period doesn't seem right, busy checking with Richard what should we use.

lauravignoli committed:
COM-89 Fix getDateOfVirologicTest

lauravignoli committed:
COM-89 Use a different function for checking if in ARV treatment

The previous function was not checking if the scheduled drug order is still valid before the virologic test.
